### PR TITLE
Details list drag drop

### PR DESCRIPTION
**Pull request by aaronyzhou**
#### Pull request checklist

- [x] Addresses an existing issue: #658 
- [x] Include a change request file if publishing 
- [x] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### Description of changes

Fix the onDrop handler on DetailsList and also take onDragEnd event. 

#### Focus areas to test

DetailsList component and also GroupList component and everything else that uses DragDropEvents.

### CODE REVIEW

File: common/changes/detailsListDragDrop_2017-03-20-17-59.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Implement Drag Drop support for DetailsList component",
```
Comment (cschleiden) [inline common/changes/detailsListDragDrop_2017-03-20-17-59.json]: Please change to `"DetailsList: Impl...`

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ import {
 } from './interfaces';
 import { ISelection } from '../../utilities/selection/interfaces';
 
-const DISTANCE_FOR_DRAG_SQUARED = 25; // the minimum mouse move distance to treat it as drag event
+const DISTANCE_FOR_DRAG_SQUARED = 9; // the minimum mouse move distance to treat it as drag event
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: isn't 3 pixel a bit low?

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx [hunk 3]
```diff
@@ -155,13 +164,20 @@ export class DragDropHelper implements IDragDropHelper {
...
       }
 
       if (isDraggable) {
         // If the target is draggable, wire up local event listeners for mouse events.
         onMouseDown = this._onMouseDown.bind(this, dragDropTarget);
+        onDragEnd = (event: DragEvent) => {
+          if (dragDropOptions.onDragEnd) {
+            dragDropOptions.onDragEnd(dragDropOptions.context.data, event);
+          }
+        };
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: why not a method on the class?

File: apps/fabric-examples/src/pages/DetailsListPage/examples/DetailsList.DragDrop.Example.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        />
+      </div>
+    );
+  }
+
+  private _getDragDropEvents(): IDragDropEvents {
+    return {
+      canDrop: (dropContext?: IDragDropContext, dragContext?: IDragDropContext) => { return true; },
+      canDrag: (item?: any) => { return true; },
+      onDragEnter: (item?: any, event?: DragEvent) => { return 'drag-enter'; }, // return string is the css classes that will be added to the enterring element.
```
Comment (cschleiden) [inline apps/fabric-examples/src/pages/DetailsListPage/examples/DetailsList.DragDrop.Example.tsx]: nit, spelling in comment

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx [hunk 5]
```diff
@@ -197,6 +197,7 @@ export class DetailsRow extends React.Component<IDetailsRowProps, IDetailsRowSta
         data-selection-index={ itemIndex }
         data-item-index={ itemIndex }
         data-is-draggable={ isDraggable }
+        draggable= { isDraggable }
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx]: what's the difference between the two?

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 3 above)
Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: What do you mean by method on the class? We're trying to use the dragEnd function on the interface we passed it.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 2 above)
Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: I thought that 5 pixels was a little too high. 16?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx (hunk 5 above)
Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx]: We're not using data-is-draggable anywhere else. Not sure if we can remove it. draggable is the html5 property to enable an element to be dragged.

File: apps/fabric-examples/src/pages/DetailsListPage/examples/DetailsList.DragDrop.Example.tsx (hunk 4 above)
Comment (aaronyzhou) [inline apps/fabric-examples/src/pages/DetailsListPage/examples/DetailsList.DragDrop.Example.tsx]: fixed

File: common/changes/detailsListDragDrop_2017-03-20-17-59.json (hunk 1 above)
Comment (aaronyzhou) [inline common/changes/detailsListDragDrop_2017-03-20-17-59.json]: fixed

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 3 above)
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: I think the expectation here is to see:
```ts
onDragEnd = this._onDragEnd;
```
Where `this._onDragEnd` has the `@autobind` decorator.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx [hunk 6]
```diff
@@ -285,8 +304,7 @@ export class DragDropHelper implements IDragDropHelper {
         let xDiff = this._dragData.clientX - event.clientX;
         let yDiff = this._dragData.clientY - event.clientY;
         if (xDiff * xDiff + yDiff * yDiff >= DISTANCE_FOR_DRAG_SQUARED) {
-          if (this._dragData.dragTarget &&
-            this._selection.isIndexSelected(options.selectionIndex)) {
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: Removing the selection check changes an expectation regarding the behavior. Right now starting a drag on an unselected item can create a selection marquee, or allow for a pan. If we remove the selection check, we'll need another way to allow the host component to control when a drag/drop operation may start.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 2 above)
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: The lower this number goes, the easier it is for there to be accidental drags. Drag/drop which results in a move or other immediate-commit operation can cause user grief.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 6 above)
Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: We already have a canDrag option in dragDropEvents and users can check for the selection in that function if they want to.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx [hunk 7]
```diff
@@ -229,6 +250,7 @@ export class DragDropHelper implements IDragDropHelper {
       if (this._dragData.dropTarget) {
         // raise dargleave event to let dropTarget know it need to remove dropping style
```
Comment (mohitbagra) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: typo in "dargleave"

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx (hunk 5 above)
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx]: @dzearing any idea who might know more about `data-is-draggable` in case it's really not used (haven't checked)?

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx (hunk 6 above)
Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: It looks like drag/drop won't work anyways since the selection object and marquee selection will steal the mousedown event. @ThomasMichon

Comment (aaronyzhou) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: Ok so we decided to revert this change. I think this should still work with your scenario with the marquee selection.

**Comment by cschleiden** [comment]
At least in the new example I find the behavior confusing: 
dnd

To me, as a casual user, the drag/drop behavior between selected/non-selected items looks exactly the same.
